pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,378 @@
+	a.SetDefault(ParamLabelTagWhitelist, DefaultLabelTagWhitelist)
+	a.SetDefault(ParamResyncPeriod, DefaultResyncPeriod)
+	a.SetDefault(ParamWatchCluster, DefaultWatchCluster)
+
+	// Create the tag whitelists
+	annotationWhitelist, err := stringListToRegexList(a.GetStringSlice(ParamAnnotationTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+	labelWhitelist, err := stringListToRegexList(a.GetStringSlice(ParamLabelTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up informers
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+
+	watchCluster := a.GetBool(ParamWatchCluster)
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+	if watchCluster != true {
+		if options.NodeName == "" {
+			return nil, errors.New(fmt.Sprintf(
+				"'%s' set to false, and node name could not be found from environment variable '%s'",
+				ParamWatchCluster, NodeNameEnvVar))
+		}
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fmt.Sprintf("spec.nodeName=%s", options.NodeName)
+		}
+	}
+
+	podsInf := core_v1inf.NewFilteredPodInformer(
+		clientset,
+		meta_v1.NamespaceAll,
+		a.GetDuration(ParamResyncPeriod),
+		indexers,
+		customWatchOptions)
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
+	// TODO: we should emit events to the k8s API to fit in with the k8s ecosystem
+
+	return &Provider{
+		ClientSet:           &clientset,
+		PodsInf:             podsInf,
+		annotationWhitelist: annotationWhitelist,
+		labelWhitelist:      labelWhitelist,
+		logger:              options.Logger,
+	}, nil
+}
+
+func (p *Provider) Run(ctx context.Context) {
+	var wg wait.Group
+	defer wg.Wait()
+
+	informerCtx, informerCancel := context.WithCancel(ctx)
+	defer informerCancel()
+
+	wg.StartWithChannel(informerCtx.Done(), p.PodsInf.Run) // Fork
+
+	// Work until asked to stop
+	<-ctx.Done()
+}
+
+func GetNodeName() string {
+	// To ensure we get this information in a timely manner this MUST be passed down to the pod via an environment
+	// variable using the downwards API
+	// See: https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/
+	return os.Getenv(NodeNameEnvVar)
+}
+
+func tagKeyMatchesWhitelist(tagKey string, whitelist []regexp.Regexp) bool {
+	for _, r := range whitelist {
+		if r.MatchString(tagKey) {
+			return true
+		}
+	}
+	return false
+}
+
+func stringListToRegexList(strList []string) ([]regexp.Regexp, error) {
+	var regexList []regexp.Regexp
+	for _, str := range strList {
+		regex, err := regexp.Compile(str)
+		if err != nil {
+			return nil, err
+		}
+		regexList = append(regexList, *regex)
+	}
+	return regexList, nil
+}
+
+func getSelfIPAddress() (gostatsd.IP, error) {
+	// In Kubernetes, the IP address of eth0 will be the IP address of the pod running this application
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return "", err
+	}
+	for _, i := range interfaces {
+		if i.Name != "eth0" {
```

Comment:
So the main other option would be to supply via the downwards API, but I tried to avoid needing that at all since it's a bit more of a pain to enforce.

We could just make the executive decision to never allow enriching our own pod metrics via this mechanism, which would probably be fine. Then we wouldn't need our own IP.

Reply:
I'm aware it's going to add additional complexity, but since we will have the node name, could we grab the IP from the node object in the cluster?

Reply:
I think it should either be via downward API (my preferred way) or via a configuration field/env var. Fetching it from `Node` objects will require adding extra permissions to `ServiceAccount`. It's nice to keep it minimal.

Reply:
Good thoughts about RBAC, I hadn't thought that far ahead.